app/src/main/java/ryey/easer/skills/operation/http_request/HttpRequestOperationData.java

```diff
@@ -0,0 +1,194 @@
+        this.requestMethod = requestMethod;
+        this.url = url;
+        this.requestHeader = requestHeader;
+        this.contentType = contentType;
+        this.postData = postData;
+    }
+
+    HttpRequestOperationData(@NonNull String postData, @NonNull PluginDataFormat format, int version) throws IllegalStorageDataException {
+        parse(postData, format, version);
+    }
+
+    public void parse(@NonNull String data, @NonNull PluginDataFormat format, int version) throws IllegalStorageDataException {
+        switch (format) {
+            default:
+                try {
+                    JSONObject jsonObject = new JSONObject(data);
+                    requestMethod = RequestMethod.valueOf(jsonObject.getString(K_REQUEST_METHOD));
+                    url = jsonObject.getString(K_URL);
+                    requestHeader = jsonObject.getString(K_REQUEST_HEADER);
+                    contentType = jsonObject.getString(K_CONTENT_TYPE);
+                    postData = jsonObject.getString(K_POST_DATA);
+                } catch (JSONException e) {
+                    e.printStackTrace();
+                    throw new IllegalStorageDataException(e);
+                }
+        }
+    }
+
+    @NonNull
+    @Override
+    public String serialize(@NonNull PluginDataFormat format) {
+        try {
+            JSONObject jsonObject = new JSONObject();
+            jsonObject.put(K_REQUEST_METHOD, requestMethod);
+            jsonObject.put(K_URL, url);
+            jsonObject.put(K_REQUEST_HEADER, requestHeader);
+            jsonObject.put(K_CONTENT_TYPE, contentType);
+            jsonObject.put(K_POST_DATA, postData);
+            return jsonObject.toString();
+        } catch (JSONException e) {
+            e.printStackTrace();
+            throw new IllegalStateException(e.getMessage());
+        }
+    }
+
+    @SuppressWarnings("RedundantIfStatement")
+    @Override
+    public boolean isValid() {
+        if (Utils.isBlank(url)) {
+            return false;
+        }
+
+        if (requestMethod == RequestMethod.POST && Utils.isBlank(contentType)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        HttpRequestOperationData that = (HttpRequestOperationData) o;
+        return requestMethod.equals(that.requestMethod) &&
+                url.equals(that.url) &&
+                requestHeader.equals(that.requestHeader) &&
+                contentType.equals(that.contentType) &&
+                postData.equals(that.postData);
+    }
+
+    @Override
+    public int hashCode() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
+            return Objects.hash(requestMethod, url, requestHeader, contentType, postData);
+        } else {
+            return requestHeader.hashCode() * 37 * 37 * 37 * 37
```

Comment:
Will this result in integer overflow?
37^4 > 2^16, so it's likely the final number may exceed 2^32.

Reply:
Yes, it can result in an integer overflow.  Event `requestHeader.hashCode() * 2` could result in one.  But for hashCodes that's actually not important if it does and it's as far as I know common practice to generate hashCode functions that way.  But I can change it.  Any suggestions?

Reply:
I replaced all 37s by 31s.